### ISSUE TITLE
SofT-GRPO scores perturbed logits, not parameter-free Gumbel-noise actions

### ISSUE THREAD

**Issue by microprediction** [microprediction/latentreasoning#66]
## Reviewed source

Default-branch commit: [`11b700b`](https://github.com/microprediction/latentreasoning/commit/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51)

Affected text:

- [`training.html` lines 237–264](https://github.com/microprediction/latentreasoning/blob/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51/training.html#L237-L264)
- [`training.html` lines 332–343](https://github.com/microprediction/latentreasoning/blob/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51/training.html#L332-L343)
- [`bibliography.html` lines 1125–1126](https://github.com/microprediction/latentreasoning/blob/11b700ba0c9cc03b0e1a54db15fe06d15fad8e51/bibliography.html#L1125-L1126)

The table is explicitly organized around “the action and its density.” Its SofT-GRPO row identifies Gumbel noise and the density of recovered noise, then labels the estimator “GRPO with Gumbel reparameterization.” The prose likewise says the policy density is the Gumbel density of the recovered noise. This conflates two different action coordinates and two different gradient estimators.

I searched open and closed issues for SofT-GRPO, Gumbel, and reparameterization. The nearest results (#4, #42, and #54) concern terminal policy gradients, Wu et al.'s top-1 diagnosis, and pass@k respectively, not this distinction.

## The parameter-free noise cannot itself supply the reported policy ratio

Write the rollout used by [Zheng et al., arXiv:2511.06411v2, §3.1–3.2](https://arxiv.org/html/2511.06411#S3) as

```text
ε_i ~ f_Gumbel                 (independent of θ)
g'_i = log p_{θ,i} + ε_i
s = Eᵀ softmax(g'/τ).
```

If `ε` is the RL action, as the paper says in §3.2 and Appendix C.3, then its policy density is

```text
π_θ(ε) = ∏_i f_Gumbel(ε_i),
```

which is independent of `θ`. Therefore

```text
∇_θ log π_θ(ε) = 0
```

and the importance ratio for the same noise action under current and old policies is identically one. Calling the base noise the action cannot produce the nonzero GRPO credit assignment shown on the page.

## The formula actually defines a policy over perturbed logits

What the paper really stores and scores is `g'`. For that random variable, the policy-induced density is the location-family density

```text
q_θ(g') = ∏_i f_Gumbel(g'_i - log p_{θ,i}).
```

The old rollout has `g' = log p_old + ε_old`, and the current-policy numerator evaluates the same stored `g'` using the residual

```text
ε_θ = g' - log p_θ.
```

That is exactly the paper's equations (11), (12), and (29), including the derivation in [Appendix C.3](https://arxiv.org/html/2511.06411#A2.SS3). It is coherent if **the perturbed-logit vector `g'` is the action** and the map `g' → s` is treated as a fixed deterministic environment transition.

The paper itself switches between the incompatible descriptions. Main-text §3.2 says “We take ε as RLVR actions” but immediately writes `p(g' | ·, θ)`; Appendix C.3 first says ε is the action and then says “we consider outputting g' as an action.”

This distinction also explains why the density is not a density of the actual soft token. The map

```text
g' → Eᵀ softmax(g'/τ)
```

is many-to-one: softmax is invariant to adding a common constant, and the vocabulary-to-embedding projection is non-injective when vocabulary size exceeds embedding dimension. The paper explicitly invokes this many-to-one map as its reason for scoring `g'` instead of `s`.

## This is a score-function estimator, not a pathwise reparameterization gradient

The displayed update is an advantage multiplied by a current/old density ratio for `q_θ(g')`. Its gradient contains `A ∇_θ log q_θ(g')`: a likelihood-ratio/score-function policy-gradient term.

A pathwise reparameterization estimator would instead hold `ε` fixed and differentiate the downstream reward or differentiable surrogate through

```text
g'(θ, ε) = log p_θ + ε.
```

These are distinct estimators. Schulman et al., [*Gradient Estimation Using Stochastic Computation Graphs*, §2.1](https://arxiv.org/html/1506.05254#S2.SS1), give the score-function formula in their equations (1)–(2) and the pathwise derivative in equation (3). SofT-GRPO uses the former inside a clipped GRPO/PPO objective; Gumbel noise provides a convenient location-family action distribution, but the reported update does not backpropagate reward through a reparameterized sample path.

## Consequence

This does **not** show that SofT-GRPO's update is invalid. A policy may legitimately sample an augmented action `g'` even when many such actions produce the same environmental input `s`. But the distinction matters to the page's central taxonomy:

- the stochastic action carrying a parameter-dependent policy density is `g'`, not the base noise `ε`;
- “recovered noise” is the residual used to evaluate the density of the stored `g'` under a candidate policy;
- the GRPO update is a score-function/likelihood-ratio estimator, not a pathwise Gumbel-reparameterization gradient;
- the formula is not a density over soft tokens.

## Concrete correction

For example, replace the row and explanatory sentence with:

> SofT-GRPO samples perturbed logits `g' = log p_θ + ε` with iid Gumbel base noise, then deterministically maps `g'` to a soft token. It treats `g'` as the augmented action and evaluates its location-family density as `q_θ(g') = ∏ f_Gumbel(g'_i - log p_{θ,i})`. The current/old ratio therefore gives a score-function GRPO/PPO update. The paper calls this “Gumbel reparameterization,” but it is not the pathwise reparameterization gradient through `g'(θ, ε)`, and it is not a density over the many-to-one soft-token image.

The bibliography can retain the paper's title while describing the estimator with this qualification.

**Comment by microprediction** [comment]
Partly correct, and the part that is correct is fixed in 1185867.

Your analysis of the mechanism checks out against the paper. Equation 11 gives a theta-free Gumbel density, Equation 12 evaluates it at the residual between the perturbed logits and the current ones, and Appendix C.3 switches from treating epsilon as the action to treating g-prime as the action. If epsilon really were the action the ratio would be identically one. It is a likelihood-ratio estimator, not a pathwise one.

Where I think the issue overstates: the prose already said the density is "recovered under the current parameters as the difference between the perturbed and unperturbed logits", which is the residual, not a parameter-free epsilon density. So two of the consequence bullets argue against a reading the page did not quite assert.

The real defect was the label. The table called it "GRPO with Gumbel reparameterization", and reparameterization invites exactly the pathwise reading the mechanism does not support. That is now "GRPO on a perturbed-logit score", with the estimator type stated in the cell.
